mon: update mgr CLI command to display detailed output for disabled modules

Acewvrs opened the pull request:
Related Trello card: https://trello.com/c/eDEORDHm/595-ceph-mgr-module-ls-should-display-detailed-output-for-enabled-modules-like-it-does-for-disabled-modules

This PR enhances the ceph mgr module ls -f json-pretty command to include detailed information about “enabled” and “always-on” modules. Previously, the command only displayed detailed information for disabled modules, and this update ensures comprehensive JSON output for all module states.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

ljflores reviewed (commented):
Please rebase the branch and fix the conflicts like we discussed during the RCOS meeting. Once you resolve conflicts, you should see the "needs-rebase" label get removed by the GitHub bot.

Acewvrs commented:
> Please rebase the branch and fix the conflicts like we discussed during the RCOS meeting. Once you resolve conflicts, you should see the "needs-rebase" label get removed by the GitHub bot.

Hi, the merge conflict has been resolved, but a different error occurred after rebasing.

ljflores commented:
Oh I see, the rebase label was removed. Good!

Let's see about the remaining test failures...
1. One of the commits still isn't signed off correctly (you can check the "details" tab in the "Signed-off-by" check.
2. The api check failed from a Python unit test. At a glance, it doesn't look related since your PR didn't make changes to the Python code. But you can inspect that further by checking the "details" tab. If you want to rerun the api check to try and get it to pass, you can do so by commenting "jenkins test api".
3. The make check test failed from a seemingly unrelated issue too:
```
[PostBuildScript] - [INFO] Executing post build scripts.
[PostBuildScript] - [INFO] Build does not have any of the results [ABORTED]. Did not execute build step #0.
INFO: Processing CTest-Version 3.x (default)
Agent went offline during the build
ERROR: Connection was broken
Build step 'Publish xUnit test result report' marked build as failure
```
Basically, if a test fails, you'll want to check if there is a compile time problem related to the part of the code that you changed. This seems like an infrastructure issue, where the node that was running the unit tests broke its connection. I would also retrigger this test to try to get it to pass with "jenkins test make check".

ljflores commented:
Also, please add a description of the change to the pull request. You can click on the three dots above "Contribution Guidelines" to edit the description.

Acewvrs commented:
jenkins test make check

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
> Oh I see, the rebase label was removed. Good!
> 
> Let's see about the remaining test failures...
> 
> 1. One of the commits still isn't signed off correctly (you can check the "details" tab in the "Signed-off-by" check.
> 2. The api check failed from a Python unit test. At a glance, it doesn't look related since your PR didn't make changes to the Python code. But you can inspect that further by checking the "details" tab. If you want to rerun the api check to try and get it to pass, you can do so by commenting "jenkins test api".
> 3. The make check test failed from a seemingly unrelated issue too:
> 
> ```
> [PostBuildScript] - [INFO] Executing post build scripts.
> [PostBuildScript] - [INFO] Build does not have any of the results [ABORTED]. Did not execute build step #0.
> INFO: Processing CTest-Version 3.x (default)
> Agent went offline during the build
> ERROR: Connection was broken
> Build step 'Publish xUnit test result report' marked build as failure
> ```
> 
> Basically, if a test fails, you'll want to check if there is a compile time problem related to the part of the code that you changed. This seems like an infrastructure issue, where the node that was running the unit tests broke its connection. I would also retrigger this test to try to get it to pass with "jenkins test make check".

Thank you! I got the make check to pass. I'll rerun the API test now.

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test make check

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test make check

ljflores reviewed (approved):
LGTM. I am going to request a few more reviews before this goes into quality assurance testing to be sure.

ronen-fr commented on src/mon/MgrMonitor.cc:
```diff
@@ -1012,13 +1012,24 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
     }
     f->flush(rdata);
-  } else if (prefix == "mgr module ls") {
+} else if (prefix == "mgr module ls") {
```
why this indentation change?

ronen-fr commented on src/mon/MgrMonitor.cc:
```diff
@@ -1012,13 +1012,24 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
+} else if (prefix == "mgr module ls") {
     if (f) {
+      // Create a mapping from module names to ModuleInfo pointers
+      std::unordered_map<std::string, const MgrMap::ModuleInfo*> module_info_map;
```
You are manipulating pointers into a vector, if I'm not mistaken. I do not know the locking infra for the modules. Are we sure there is no possibility of change to the modules vector by another thread during the execution of this function?

rzarzynski replied:
Yeah. How about dropping the intermediary container and use e.g. `std::find_if` instead?

Acewvrs commented on src/mon/MgrMonitor.cc:
```diff
@@ -1012,13 +1012,24 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
+
       f->open_object_section("modules");
       {
+        // always_on_modules
```
> If you need a block - better just name it (as an aux function)

Would it be okay if I just remove the block, instead of creating a whole new function?

rzarzynski commented on qa/tasks/mgr/mgr_test_case.py:
```diff
@@ -109,9 +109,7 @@ def setup_mgrs(cls):
         module_names = {module['name'] for module in loaded}
-        print("module names: ", module_names)
         unload_modules = module_names - {"cephadm"}
-        print("unload modules: ", unload_modules)
```
Please drop this and previous commit – together they form _no change_.

rzarzynski reviewed (commented):
Thank you for bringing this! Please see the comments below.

Acewvrs commented:
jenkins test windows

Jordan231111 commented:
jenkins test make check arm64

Acewvrs commented:
jenkins test make check

Acewvrs commented:
jenkins test make check arm64

Acewvrs commented:
jenkins test make check

Acewvrs commented:
jenkins test make check arm64

Acewvrs commented:
jenkins test make check

Acewvrs commented:
jenkins test make check

ljflores reviewed (commented):
Hi @Acewvrs, the logic looks good to me. Can you take one more step to clean up the commits and squash them into one or a few distinct logical changes? For instance, "mon: fix spacing" can be grouped in with another commit since this is an after-cleanup.

I have also re-requested the reviews of those who dropped comments before in case there are any final changes to be made. Then, this can go into QA!

ljflores commented:
If you are unable to complete the PR, let me know and I can take it over.

Acewvrs commented:
Hi @ljflores, I've squashed the commits and grouped them into 3 main categories:

1. Our initial approach to the problem using a hash map.
2. Fixing the test to accommodate the changes we made from group 1. 
3. Our more refined solution, which reduces memory usage by removing the hash map.

I've added brief descriptions to each commit as well.

ronen-fr commented on src/mon/MgrMonitor.cc:
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
-        f->close_section();
+      
+      // always_on_modules
+      f->open_array_section("always_on_modules");
```
(not a must for this PR, but if you feel like it:
you can change - in a separate commit - the way sections are opened and closed in the new code to
use the safer RAII wrappers. See for example https://github.com/ceph/ceph/pull/61618 for details).

ronen-fr commented on src/mon/MgrMonitor.cc:
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
     f->flush(rdata);
   } else if (prefix == "mgr module ls") {
     if (f) {
+      const auto always_on_modules = map.get_always_on_modules();
```
please don't reuse the exact name of the map member. It's confusing.

ronen-fr commented on src/mon/MgrMonitor.cc:
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
+        if (always_on_modules.count(module_name) > 0)
+          continue;
+        auto module_itr = std::find_if(
+          available_modules.begin(), 
```
I am missing something...  isn't your available_modules the set that you have created earlier? a set of strings (the module names)? if so - why find_if(), and not just find()?

rzarzynski replied:
My understanding is that:

1. `available_modules` is as below and stores `MgrMap::ModuleInfo` instances:
```cpp
 const auto available_modules = map.available_modules;
```
2. The `find_if` is for the sake of accessing the `name` member of `MgrMap::ModuleInfo`.

ronen-fr commented on src/mon/MgrMonitor.cc:
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
+      // disabled_modules
+      f->open_array_section("disabled_modules");
+      for (const auto& p : map.available_modules) {
+        if (map.modules.count(p.name) == 0 &&
```
contains

Acewvrs commented:
jenkins test api

rzarzynski commented on src/mon/MgrMonitor.cc:
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
+      // always_on_modules
+      f->open_array_section("always_on_modules");
+      for (const auto& module_name : always_on_modules) {
+        auto module_itr = std::find_if(
```
Would be nice!

ljflores commented:
Hey @Acewvrs, IMO the 4 commits can be squashed into just one commit. Can you squash it, and consolidate the explanations into one description?

@ronen-fr Can you take a final look at this before it enters QA?

Acewvrs commented:
> Hey @Acewvrs, IMO the 4 commits can be squashed into just one commit. Can you squash it, and consolidate the explanations into one description?

I’ve squashed the 4 commits into a single commit.

ronen-fr reviewed (approved):
LGTM.
One suggested improvement: use the RAII "dump sections" (see https://github.com/ceph/ceph/pull/61506). But this can be done in
a followup PR.

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test make check

Acewvrs commented:
Jenkins test api

Acewvrs commented:
Jenkins test api

Acewvrs commented:
Jenkins test api

Acewvrs commented:
Jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

Acewvrs commented:
jenkins test api

ceph-jenkins commented:
Can one of the admins verify this patch?
